Provide API to set/get the number of I/O threads

kingcrimsontianyu opened the pull request:
## Description

This PR allows users to programmatically set the number of I/O threads in C++ and Python. The Python part is implemented using [cuDF's option system](https://docs.rapids.ai/api/cudf/stable/user_guide/api_docs/options/).

Partially addresses #18272

<!-- Provide a standalone description of changes in this PR. -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
<!-- Note: The pull request title will be included in the CHANGELOG. -->

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

kingcrimsontianyu commented:
## Sample code

```python
import cudf
import cupy
import rmm

rmm.mr.set_current_device_resource(rmm.mr.CudaAsyncMemoryResource())
cupy.random.seed(2077)
df = cudf.DataFrame({"a": cupy.random.rand(1 * 10**8)})
filepath = "tmp.pq"
df.to_parquet(filepath)

# Use the default number of IO threads (4 by our current implementation)
df = cudf.read_parquet(filepath)

# Change the setting globally
cudf.set_option("num_io_threads", 8)

# Use 8 threads now
df = cudf.read_parquet(filepath)

# Use 2 threads specified via the context manager
with cudf.option_context("num_io_threads", 2):
    df = cudf.read_parquet(filepath)

# Use 8 threads again
df = cudf.read_parquet(filepath)
```

kingcrimsontianyu commented:
/ok to test 789fff7e97488eeb306e6e7121e11ee0d862fde3

Matt711 commented on python/pylibcudf/pylibcudf/io/kvikio_manager.pyx:
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.
```
Add a type stub file `kvikio_manager.pyi` with
```python
# Copyright (c) 2025, NVIDIA CORPORATION.

def set_num_io_threads(num_io_threads: int) -> int: ...

def get_num_io_threads() -> int: ...

def get_default_num_io_threads() -> int: ...
```

kingcrimsontianyu replied:
Thanks for the suggestion. Added the stub file.

I'm not quite sure if the `unsigned int` type needs any special handling in the stub file, though. At the moment, I'm leaving it as `int` as suggested.

Matt711 replied:
No I thinks it's the best we can do.
> For typed integer arguments in libcudf, use int as a type annotation.

https://docs.rapids.ai/api/cudf/stable/pylibcudf/developer_docs/#type-stubs

Matt711 reviewed (commented):
Thanks @kingcrimsontianyu, mostly non-blocking comments. However, I did want to know what you and others think about changing the option name.

vuule commented on python/cudf/cudf/tests/test_options.py:
```diff
@@ -127,3 +127,20 @@ def test_options_context_error():
+    new_setting_1 = 16
+    new_setting_2 = 2
+    new_setting_3 = 20
+    with cudf.option_context("num_io_threads", new_setting_1):
```
could it be just `io.num_threads`? I'm not sure if users should to be aware of kvikio, if we can help it. I'm fine with `io.kvikio.num_threads` if you want to be transparent about kvikio use.

kingcrimsontianyu replied:
Done. Used `io.num_threads` for the time being for brevity.

Matt711 replied:
I'm happy with `io.num_threads`

vuule commented on cpp/tests/io/kvikio_manager_test.cpp:
```diff
@@ -0,0 +1,42 @@
+  // Use a new setting
+  {
+    unsigned int new_setting{8u};
+    cudf::io::kvikio_manager::set_num_io_threads(new_setting);
```
I think we also need to assert that the new setting is actually different from the default

kingcrimsontianyu replied:
We will have a platform-dependent default in a future PR that may equal the new setting.

vuule replied:
then we should try to pick a value here such that it's very unlikely to match the default

kingcrimsontianyu replied:
I can pick an odd number for that purpose. 

On a second thought, it is probably not a good idea to let a processor-dependent variable participate in the unit test. To maintain consistency, I've removed that part.

vuule commented on cpp/src/io/utilities/data_sink.cpp:
```diff
@@ -33,7 +34,7 @@ class file_sink : public data_sink {
   explicit file_sink(std::string const& filepath)
   {
-    kvikio_integration::set_up_kvikio();
+    kvikio_manager::instance();
```
maybe a comment on why this is done?

kingcrimsontianyu replied:
Done.

vuule commented on cpp/src/io/utilities/kvikio_manager.cpp:
```diff
@@ -0,0 +1,96 @@
+
+#include <cuda_runtime.h>
+
+namespace cudf {
+namespace io {
```
```suggestion
namespace cudf::io {
```

kingcrimsontianyu replied:
Done.

vuule commented on cpp/src/io/utilities/kvikio_manager.cpp:
```diff
@@ -0,0 +1,96 @@
+    "Set the number of I/O threads. Old value: %u, new value: %u", old_setting, num_io_threads);
+}
+
+unsigned int kvikio_manager::get_num_io_threads() { return instance()._num_io_threads; }
+
+unsigned int kvikio_manager::get_default_num_io_threads()
```
[[nodiscard]] these

kingcrimsontianyu replied:
Done.

vuule commented on cpp/src/io/utilities/kvikio_manager.cpp:
no need for the get_ prefixes

kingcrimsontianyu replied:
Done. I searched `get_prefixes` but found nothing, and then I saw the white space.

kingcrimsontianyu commented on cpp/include/cudf/io/kvikio_manager.hpp:
```diff
@@ -0,0 +1,93 @@
+   * If the new value is different from the previous value, then this method will block the calling
+   * thread until KvikIO completes all existing I/O tasks, destroys the previous thread pool, and
+   * creates a new one with the specified value. Otherwise, the existing thread pool will be used
+   * for subsequent I/O operations.
```
Thanks. I've itemized the description to further improve clarity.

kingcrimsontianyu commented on cpp/include/cudf/io/kvikio_manager.hpp:
```diff
@@ -0,0 +1,93 @@
+   *
+   * @note This function instantiates the kvikio_manager if it does not exist.
+   */
+  static void set_num_io_threads(unsigned int num_io_threads);
```
If the number is 0, KvikIO will throw an exception and ask for a positive integer. But we do need to add extra sanity check here. Done as suggested.

kingcrimsontianyu commented on cpp/src/io/utilities/kvikio_manager.cpp:
```diff
@@ -0,0 +1,96 @@
+    kvikio::defaults::set_thread_pool_nthreads(num_io_threads);
+  }
+  CUDF_LOG_INFO(
+    "Set the number of I/O threads. Old value: %u, new value: %u", old_setting, num_io_threads);
```
Thanks. I intend to log all the setter events.

To improve clarity, I have made the log message different on the two different branches.

kingcrimsontianyu commented on python/cudf/cudf/options.py:
```diff
@@ -366,6 +389,43 @@ def _integer_and_none_validator(val):
+def _num_io_threads_get_callback():
+    actual_result = plc.io.kvikio_manager.get_num_io_threads()
+    expected_result = _OPTIONS["num_io_threads"].value
+    assert actual_result == expected_result
```
Added the assertion message as suggested.

Yes. We want to raise an exception if there's a mismatch. These two values must always be equal. Otherwise I think we may be having more serious problems such as memory corruption (flipped bits causing the values stored in Python `_OPTIONS` and C++ `_num_io_threads` to differ) or race condition (setting the option from different threads without a lock).

kingcrimsontianyu commented on python/cudf/cudf/options.py:
```diff
@@ -366,6 +389,43 @@ def _integer_and_none_validator(val):
+        There are different ways to set this value. In descending order of
+        override priority:
+            - cuDF option API
+                - cudf.set_option("num_io_threads", value) to set the value
```
Done. Thanks! I always conflate the different ways to treat the sample code in Sphinx, Doxygen, and Markdown.

kingcrimsontianyu commented on python/cudf/cudf/options.py:
```diff
@@ -366,6 +389,43 @@ def _integer_and_none_validator(val):
+                - cudf.set_option("num_io_threads", value) to set the value
+                globally
+                - cudf.option_context("num_io_threads", value) to set the value
+                in a context manager
```
Done.

kingcrimsontianyu commented on python/cudf/cudf/options.py:
```diff
@@ -366,6 +389,43 @@ def _integer_and_none_validator(val):
+                globally
+                - cudf.option_context("num_io_threads", value) to set the value
+                in a context manager
+            - Environment variable KVIKIO_NTHREADS
```
Done.

kingcrimsontianyu commented on python/cudf/cudf/options.py:
```diff
@@ -366,6 +389,43 @@ def _integer_and_none_validator(val):
+            - cuDF's default number of IO threads
+                cuDF derives a default platform-dependent value, which
+                can be queried by the method
+                pylibcudf.io.kvikio_manager.get_default_num_io_threads().
```
Done.

kingcrimsontianyu commented on python/cudf/cudf/options.py:
```diff
@@ -366,6 +389,43 @@ def _integer_and_none_validator(val):
+        \tValid values are integers. Default is a platform-dependent value.
+    """
+    ),
+    _integer_validator,
```
Done. Added an extra validator.

kingcrimsontianyu commented:
Is there a way to rebase the branch onto dev, but still allow reviewers to see "the changes since last review"? I need the rebase because it will fix the CI failure.

vuule reviewed (commented):
few nitpicks

vyasr commented:
> Is there a way to rebase the branch onto dev, but still allow reviewers to see "the changes since last review"? I need the rebase because it will fix the CI failure. Will do a merge, but apologies in advance if this causes chaos in review.

I don't follow, but I suspect that you are confusing the effects of merging and rebasing. A merge is equivalent to a rebase for the purposes of fixing the CI failure, but only rebases are a problem for reviews. Rebases rewrite the history of the repository and Github attaches all review comments to particular commit hashes, so if you rebase and force push the history is mangled and the data is lost. A merge results in a cleaner review process and is totally fine. Therefore, there is no reason not to merge upstream; our policy is to avoid rebasing once a PR is out of draft (or at least, once any reviews have started).

kingcrimsontianyu commented on cpp/include/cudf/io/kvikio_manager.hpp:
```diff
@@ -0,0 +1,110 @@
+   *
+   * @param num_io_threads The number of IO threads to be used.
+   *
+   * @note This function instantiates the kvikio_manager if it does not exist.
```
Done.

kingcrimsontianyu commented on cpp/include/cudf/io/kvikio_manager.hpp:
```diff
@@ -0,0 +1,110 @@
+   *
+   * @return The default number of IO threads derived by cuDF.
+   *
+   * @note This function does not instantiate the kvikio_manager if it does not exist.
```
Done.

kingcrimsontianyu commented on cpp/include/cudf/io/kvikio_manager.hpp:
```diff
@@ -0,0 +1,110 @@
+
+ private:
+  /**
+   * @brief Constructor of kvikio_manager.
```
Done.

vuule commented on cpp/src/io/utilities/kvikio_manager.cpp:
```diff
@@ -0,0 +1,120 @@
+  return kvikio_manager_instance;
+}
+
+void kvikio_manager::set_num_io_threads(unsigned int num_io_threads)
```
Reconsidered this a bit; I think it's fine as-is, unless we actually measure performance issues because of the slight redundancy here.

Matt711 reviewed (approved):
Thanks! I don't have anything else

bdice commented on cpp/include/cudf/io/kvikio_manager.hpp:
```diff
@@ -0,0 +1,110 @@
+/**
+ * @brief Utility class for data source and sink.
+ *
+ * Use this class as a data member to create a CUDA context if it has not existed, and create a
```
```suggestion
 * Use this class as a data member to create a CUDA context if no context exists, and create a
```

bdice commented on cpp/include/cudf/io/kvikio_manager.hpp:
```diff
@@ -0,0 +1,110 @@
+  [[nodiscard]] static unsigned int num_io_threads();
+
+  /**
+   * @brief Get the default number of IO threads derived by cuDF.
+   *
+   * @return The default number of IO threads derived by cuDF.
```
I'm unsure what "derived by" means in this context. Maybe this is clearer:
```suggestion
   * @brief Get the default number of IO threads used by cuDF.
   *
   * @return The default number of IO threads used by cuDF.
```

bdice commented on cpp/src/io/utilities/kvikio_manager.cpp:
```diff
@@ -0,0 +1,120 @@
+    // Workaround for https://github.com/rapidsai/cudf/issues/14140, where cuFileDriverOpen errors
+    // out if no CUDA calls have been made before it. This is a no-op if the CUDA context is already
+    // initialized.
+    cudaFree(nullptr);
```
This initialization should happen in kvikio rather than cudf, imo. This seems like an important behavior for all kvikio users and not just cudf, right?

Also, I don't like using an initializer class as a data member. The previous design where an idempotent function was called, like `kvikio_integration::set_up_kvikio();`, is a better design, because there is no state involved and the data member is never accessed. It should be thought of more like a scoped lock, where you create an object and let RAII handle things _in a local scope_. Is there a problem with doing that?

bdice commented on cpp/src/io/utilities/kvikio_manager.cpp:
```diff
@@ -0,0 +1,120 @@
+  auto const compat_mode = kvikio::getenv_or("KVIKIO_COMPAT_MODE", kvikio::CompatMode::ON);
+  kvikio::defaults::set_compat_mode(compat_mode);
+
+  std::string const env_var_name{"KVIKIO_NTHREADS"};
```
Environment variables named `KVIKIO_*` should be read and handled by KvikIO, not cuDF. We should either rename this or move the logic to KvikIO.

kingcrimsontianyu replied:
Thanks. This is new to me, and is different from the previous design.
cc @vuule on having new env vars `CUDF_*`.

bdice commented on python/cudf/cudf/options.py:
```diff
@@ -366,6 +396,45 @@ def _integer_and_none_validator(val):
+    plc.io.kvikio_manager.set_num_io_threads(_OPTIONS["io.num_threads"].value)
+
+
+def _num_io_threads_get_callback():
+    actual_result = plc.io.kvikio_manager.num_io_threads()
+    expected_result = _OPTIONS["io.num_threads"].value
+    assert actual_result == expected_result, (
+        f"Mismatch in num_io_threads: Expected: {expected_result}, Actual: {actual_result}"
+    )
```
This seems like it belongs in tests rather than library code. Even if we keep the "set callback", I don't think we should have a "get callback."

And this shouldn't raise an error -- it should treat `plc.io.kvikio_manager.num_io_threads()` as a source of truth and update `_OPTIONS` accordingly.

bdice commented on python/cudf/cudf/options.py:
```diff
@@ -58,6 +67,10 @@ def _register_option(
     validator : Callable
         Called on the option value to check its validity. Should raise an
         error if the value is invalid.
+    set_callback : Callable | None
+        Called when setting the option value.
+    get_callback : Callable | None
+        Called when getting the option value.
```
Do we really need `set_callback` and `get_callback` here?

We have other places where we control behavior, especially global state, outside of cudf Python options. Examples include the RMM memory resource and numba options. It is awkward for a cuDF option to alter the global state of another library (KvikIO). I would advocate for users to call directly through pylibcudf if this setting is specific to libcudf, or call through KvikIO Python APIs if they want to alter some global state in kvikio.

bdice replied:
Wouldn't it be better if KvikIO's Python API could be used to manage KvikIO options?

cuDF's options system shouldn't be the way that users modify the behavior of pylibcudf or cudf-polars.

vuule replied:
The challenge with directly using kvikIO APIs to set the global state is that libcudf is doing the same, so the final state depends on the order of operations.

But I think I do see the incongruity in the API design. Perhaps rebranding this as "IO settings" would make it a better fit in libcudf. kvikIO becomes an implementation detail, and the library that uses libcudf does not need to know how the IO is performed.

bdice reviewed (changes requested):
I'd like to reconsider some of the design here. We're drawing very unclear boundaries between cuDF and KvikIO as well as between libcudf/pylibcudf/cudf, which harms understandability and makes it hard to understand the structure for new readers (or infrequent I/O contributors).
